lib/wikilink.coffee

```diff
@@ -56,8 +65,27 @@ module.exports = atomWikiLink =
 
   follow: ->
     editor = atom.workspace.getActiveTextEditor()
-    [projPath, relPath] = atom.project.relativizePath(editor?.buffer?.file?.path)
-    noteDirectory = fs.normalize(projPath)
+    if not editor
+      console.log('An editor window must be active before you can follow a link.')
+      return
+
+    filePath = editor.buffer?.file?.path
+    if not filePath
+      console.log('You must save the file before you can follow a link.')
+      return
+
+    linkResolution = atom.config.get('wikilink.linkresolution')
+
+    if linkResolution is 'project'
+      [projPath, relPath] = atom.project.relativizePath(filePath)
+      if not projPath
+        console.log('The file is not saved in a project folder.')
+        return
+
+      noteDirectory = fs.normalize(projPath)
+    else if linkResolution is 'file'
+      noteDirectory = path.dirname(filePath)
+
```

Comment:
I'd like to suggest to refactor to extract this into a helper function that resolves (and returns) the `noteDirectory` or null. What do you think?

Reply:
Thanks for the review.

I tried factoring out the code, but it's a bit tricky because each of the branches does something different:

follow link:
* project resolution: Uses project path, normalizes the directory, logs on failure
* file resolution: Uses dirname

copy link:
* project resolution: Uses file's relative path, does not normalize the directory, logs on failure
* file resolution: Uses basename

get suggestions:
* project resolution: Uses project path, normalizes the directory, (silently) resolves promise on failure
* file resolution: Uses dirname


Maybe we can revisit this in the future if we end up making bug fixes to the directory code in multiple spots?

Reply:
Fine with me 👍